File: app/src/main/java/com/example/purepawapp/MainActivity.java
package com.example.purepawapp;

import android.os.Bundle;
import android.view.View;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.navigation.NavDestination;
import androidx.navigation.fragment.NavHostFragment;
import androidx.navigation.ui.NavigationUI;

import com.example.purepawapp.di.ServiceLocator;
import com.example.purepawapp.ui.common.LoadingHost;
import com.google.android.material.badge.BadgeDrawable;
import com.google.android.material.bottomnavigation.BottomNavigationView;

import java.util.Set;

public class MainActivity extends AppCompatActivity implements LoadingHost {

    private final Set<Integer> topLevelGraphIds = Set.of(
            R.id.home_nav_graph,
            R.id.product_nav_graph,
            R.id.spa_nav_graph,
            R.id.cart_nav_graph,
            R.id.account_nav_graph
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);

        CoordinatorLayout rootView = findViewById(R.id.root_coordinator);
        ViewCompat.setOnApplyWindowInsetsListener(rootView, (v, insets) -> {
            var systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, 0, systemBars.right, 0);
            return insets;
        });

        NavHostFragment navHostFragment = (NavHostFragment) getSupportFragmentManager()
                .findFragmentById(R.id.nav_host_fragment);
        var navController = navHostFragment.getNavController();

        BottomNavigationView bottomNav = findViewById(R.id.bottom_nav_view);
        NavigationUI.setupWithNavController(bottomNav, navController);

        navController.addOnDestinationChangedListener((controller, destination, arguments) -> {
            NavDestination parent = destination.getParent();
            boolean showBottomNav = (parent != null && topLevelGraphIds.contains(parent.getId()))
                    || destination.getId() == R.id.productDetailFragment;
            bottomNav.setVisibility(showBottomNav ? View.VISIBLE : View.GONE);
        });

        ServiceLocator.getCartRepository().getItemsLiveData().observe(this, items -> {
            int count = 0;
            for (var item : items) count += item.getQuantity();
            BadgeDrawable badge = bottomNav.getOrCreateBadge(R.id.cart_nav_graph);
            badge.setVisible(count > 0);
            badge.setNumber(count);
        });
    }

    @Override
    public void showLoading() {
        View overlay = findViewById(R.id.loading_overlay);
        if (overlay != null) overlay.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        View overlay = findViewById(R.id.loading_overlay);
        if (overlay != null) overlay.setVisibility(View.GONE);
    }
}

File: app/src/main/java/com/example/purepawapp/MainActivity.kt
package com.example.purepawapp

import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.setupWithNavController
import com.example.purepawapp.di.ServiceLocator
import com.google.android.material.bottomnavigation.BottomNavigationView
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    private val topLevelGraphIds = setOf(
        R.id.home_nav_graph,
        R.id.product_nav_graph,
        R.id.spa_nav_graph,
        R.id.cart_nav_graph,
        R.id.account_nav_graph
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)

        val rootView = findViewById<androidx.coordinatorlayout.widget.CoordinatorLayout>(R.id.root_coordinator)
        ViewCompat.setOnApplyWindowInsetsListener(rootView) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, 0, systemBars.right, 0)
            insets
        }

        val navHostFragment = supportFragmentManager
            .findFragmentById(R.id.nav_host_fragment) as NavHostFragment
        val navController = navHostFragment.navController

        val bottomNav = findViewById<BottomNavigationView>(R.id.bottom_nav_view)
        bottomNav.setupWithNavController(navController)

        navController.addOnDestinationChangedListener { _, destination, _ ->
            val parentId = destination.parent?.id
            val showBottomNav = parentId in topLevelGraphIds || destination.id == R.id.productDetailFragment
            bottomNav.visibility = if (showBottomNav) {
                android.view.View.VISIBLE
            } else {
                android.view.View.GONE
            }
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                ServiceLocator.cartRepository.items.collect { items ->
                    val count = items.sumOf { it.quantity }
                    val badge = bottomNav.getOrCreateBadge(R.id.cart_nav_graph)
                    badge.isVisible = count > 0
                    badge.number = count
                }
            }
        }
    }
}
